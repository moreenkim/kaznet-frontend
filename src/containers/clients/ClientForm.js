import React, { Component } from 'react';
import { Formik } from 'formik';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Redirect, Link } from 'react-router-dom';
import {
  Alert,
  Form,
  FormGroup,
  Col,
  Input,
  Button,
  Modal,
  ModalFooter,
  ModalHeader
} from 'reactstrap';

import * as errorHandlerSelectors from '../../store/errorHandler/reducer';

const transformMyApiErrors = function(array) {
  const errors = {};
  for (let index = 0; index < array.length; index++) {
    const element = array[index];
    const msg = element.detail;
    const field = element.source.pointer.split('/').pop();
    errors[field] = msg;
  }

  return errors;
};

export class ClientForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modal: false
    };
    this.toggle = this.toggle.bind(this);
    this.targetId = props.targetId || null;
  }

  toggle() {
    this.setState({
      modal: !this.state.modal
    });
  }

  render() {
    return (
      <Formik
        initialValues={{
          name: this.props.initialData.name != null ? this.props.initialData.name : null
        }}
        onSubmit={(values, { setSubmitting, setErrors, setStatus }) => {
          const payload = {
            data: {
              type: 'Client',
              id: this.targetId != null ? this.targetId : null,
              attributes: {
                name: values.name
              }
            }
          };

          try {
            this.props.formActionDispatch(payload, this.targetId).then(() => {
              setSubmitting(false);
              if (this.props.hasError) {
                setErrors(transformMyApiErrors(this.props.errorMessage));
              } else {
                setStatus('done');
              }
            });
          } catch (error) {
            console.error(error);
          }
        }}
        render={({
          values,
          errors,
          status,
          handleChange,
          handleBlur,
          handleSubmit,
          isSubmitting,
          setStatus
        }) => (
          <div>
            {errors.data && <Alert color="danger">{errors.data}</Alert>}
            <Form onSubmit={handleSubmit}>
              <FormGroup className="row">
                <Col sm="12">
                  <Input
                    name="name"
                    type="text"
                    bsSize="lg"
                    placeholder="Client Name"
                    aria-label="name"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.name}
                    className={errors.name ? 'is-invalid' : ''}
                  />
                  {errors.name && <div className="invalid-feedback">{errors.name}</div>}
                </Col>
              </FormGroup>
              <FormGroup className="row mt-5">
                <Col md={{ size: 5, offset: 1 }}>
                  <Button
                    className="btn btn-secondary btn-block"
                    aria-label="Cancel"
                    onClick={() => {
                      setStatus('done');
                    }}
                  >
                    {' '}
                    Cancel{' '}
                  </Button>
                </Col>
                <Col md={{ size: 5 }}>
                  <Button
                    type="submit"
                    className="btn btn-primary btn-block"
                    aria-label="Save Client"
                    disabled={isSubmitting}
                  >
                    {isSubmitting ? 'Saving' : 'Save Client'}
                  </Button>
                </Col>
              </FormGroup>
            </Form>
            {this.targetId != null && (
              <div>
                <hr className="my-4" />
                <Col md={{ size: 4, offset: 4 }}>
                  <Button
                    className="btn btn-secondary btn-block btn-danger center-block"
                    aria-label="DELETE CLIENT"
                    onClick={this.toggle}
                  >
                    DELETE CLIENT
                  </Button>
                </Col>
                <Modal
                  isOpen={this.state.modal}
                  toggle={this.toggle}
                  className={this.props.className}
                >
                  <ModalHeader toggle={this.toggle}>
                    Are you sure you want to delete this client?
                  </ModalHeader>
                  <ModalFooter>
                    <Link
                      to={`/clients/${this.targetId}/delete`}
                      className="btn btn-danger"
                      onClick={this.toggle}
                    >
                      Delete Client
                    </Link>
                    <Button color="secondary" onClick={this.toggle} aria-label="Cancel">
                      Cancel
                    </Button>
                  </ModalFooter>
                </Modal>
              </div>
            )}
            {status === 'done' && <Redirect to="/clients" />}
          </div>
        )}
      />
    );
  }
}

function mapStateToProps(state, ownProps) {
  return {
    hasError: errorHandlerSelectors.getHasError(state),
    errorMessage: errorHandlerSelectors.getErrorMessage(state)
  };
}

function mapDisptachToProps(dispatch, ownProps) {
  return bindActionCreators(
    {
      formActionDispatch: ownProps.action
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDisptachToProps
)(ClientForm);
