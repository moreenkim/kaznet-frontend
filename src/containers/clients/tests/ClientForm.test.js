// Test ClientForm
import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { ClientForm } from "../ClientForm";

describe("containers/task/ClientForm", () => {
  it("renders without crashing", () => {
    const initialData = { name: "Hi" };
    shallow(
      <ClientForm
        formActionDispatch={function() {}}
        initialData={initialData}
      />
    );
  });

  it("renders client form correctly", () => {
    const initialData = { name: "Hi" };
    const wrapper = mount(
      <ClientForm
        formActionDispatch={function() {}}
        initialData={initialData}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
