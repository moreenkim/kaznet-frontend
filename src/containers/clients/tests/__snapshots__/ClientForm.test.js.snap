// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/client/ClientForm renders client form correctly 1`] = `
<ClientForm
  formActionDispatch={[Function]}
  initialData={
    Object {
      "name": "Hi",
    }
  }
>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "name": "Hi",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <div>
      <Form
        onSubmit={[Function]}
        tag="form"
      >
        <form
          className=""
          onSubmit={[Function]}
        >
          <FormGroup
            className="row"
            tag="div"
          >
            <div
              className="row form-group"
            >
              <Col
                sm="12"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="col-sm-12"
                >
                  <Input
                    aria-label="name"
                    bsSize="lg"
                    className=""
                    name="name"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="Client Name"
                    type="text"
                    value="Hi"
                  >
                    <input
                      aria-label="name"
                      className="form-control-lg form-control"
                      name="name"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder="Client Name"
                      type="text"
                      value="Hi"
                    />
                  </Input>
                </div>
              </Col>
            </div>
          </FormGroup>
          <Button
            className="btn btn-primary btn-block"
            color="secondary"
            disabled={false}
            tag="button"
            type="submit"
          >
            <button
              className="btn btn-primary btn-block btn btn-secondary"
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              Save Client
            </button>
          </Button>
        </form>
      </Form>
    </div>
  </Formik>
</ClientForm>
`;
